Fix #102: Better main class detection

smarter opened the pull request:
Previously, the main class detection was handled by
https://github.com/sbt/zinc/blob/1.0/internal/zinc-apiinfo/src/main/scala/xsbt/api/Discovery.scala
which looks for a main method with the correct signature in the
extracted API. This is imperfect because it relies on ExtractAPI
dealiasing types (because Discovery will look for a main method with a
parameter type of `java.lang.String` and won't recognize
`scala.Predef.String`), dealiasing means that the extracted API looses
information and thus can lead to undercompilation.

This commit partially fixes this by adding a new callback to AnalysisCallback:
```java
    void mainClass(File sourceFile, String className)
```
that is used to explicitly register main entry points. This way, tools
do not need to interpret the extracted API, this is much better since it
makes it easier for zinc to evolve the API representation.

This commit does not actually changes ExtractAPI to not dealias, this
can be done in a later PR.

Note that there is another usecase for xsbt.api.Discovery that this PR
does not replace: discovering tests. This is more complicated because
different test frameworks have different ways to discover tests. For
more information, grep for "Fingerprint" in https://github.com/sbt/sbt
and https://github.com/sbt/junit-interface

Note also that the added scripted test does not actually test that this
PR is correct since sbt itself needs to be updated to use this new API,
the branch at
https://github.com/smarter/sbt/commits/fix/main-class-detection does
this, but cannot be merged in sbt until a new zinc is published.

Fix #102

jvican commented:
> This commit does not actually changes ExtractAPI to not dealias, this
can be done in a later PR.
  
I'm unsure this is possible at all. API detection needs dealiased types to work correctly.
  
@smarter I'll review this one, thanks for the PR.

smarter commented:
> I'm unsure this is possible at all. API detection needs dealiased types to work correctly.

API detection of what?

jvican commented:
> API detection.
  
`ExtractAPI.scala`.

smarter commented:
Yes, but why do you say that ExtractAPI needs it to work correctly? After https://github.com/sbt/zinc/pull/87 and https://github.com/sbt/zinc/pull/239 zinc should handle abstract types in the API correctly, so not expanding type aliases should work too.

smarter commented:
I fixed the added test to:
- actually work (I forgot that zinc scripted tests cannot define commands in build.sbt anymore, you have to add them to `IncHandler` instead)
- actually test zinc itself, no matter what sbt uses for main class discovery (I added a `checkMainClasses` command to `IncHandler`, that behaves similarly to existing `check*` commands in there and uses the internal API).

smarter commented:
Ping for review.

jvican commented on zinc/src/sbt-test/apiinfo/main-discovery/test:
```diff
@@ -0,0 +1,4 @@
+> compile
```
These tests look great!

jvican commented on internal/zinc-apiinfo/src/main/scala/sbt/internal/inc/ClassToAPI.scala:
```diff
@@ -94,6 +96,16 @@ object ClassToAPI {
       val defsEmptyMembers = clsDef :: statDef :: Nil
       cmap.memo(name) = defsEmptyMembers
       cmap.allNonLocalClasses ++= defs
+
```
👍

jvican reviewed (approved):
LGTM. Excellent job!
  
It's true that we need to put the hooks (or provide the necessary information) to detect main classes for tests. I'll look into the best way to do so.

jvican commented:
@smarter Sorry, but we merged a formatting PR that invalidates the contents of this one. Can you please update this PR?
  
@eed3si9n @dwijnand  Can you have a quick look to this? I want to merge it. If you don't have time, tell me so and I'll merge it.

eed3si9n reviewed (changes requested):
> This is imperfect because it relies on ExtractAPI
dealiasing types (because Discovery will look for a main method with a
parameter type of java.lang.String and won't recognize
scala.Predef.String), dealiasing means that the extracted API looses
information and thus can lead to undercompilation.

Is it possible to fix the root cause rather than adding special case for main class detection?

smarter commented:
> Is it possible to fix the root cause rather than adding special case for main class detection?

The root cause is the use of Discovery, the fix is to not use Discovery and the same should be done for tests, and Discovery should be removed. There is no way to make Discovery understand all the semantics of Scala type system.

smarter committed:
Fix #102: Better main class detection

Previously, the main class detection was handled by
https://github.com/sbt/zinc/blob/1.0/internal/zinc-apiinfo/src/main/scala/xsbt/api/Discovery.scala
which looks for a main method with the correct signature in the
extracted API. This is imperfect because it relies on ExtractAPI
dealiasing types (because Discovery will look for a main method with a
parameter type of `java.lang.String` and won't recognize
`scala.Predef.String`), dealiasing means that the extracted API looses
information and thus can lead to undercompilation.

This commit partially fixes this by adding a new callback to AnalysisCallback:
    void mainClass(File sourceFile, String className)
that is used to explicitly register main entry points. This way, tools
do not need to interpret the extracted API, this is much better since it
makes it easier for zinc to evolve the API representation.

This commit does not actually changes ExtractAPI to not dealias, this
can be done in a later PR.

Note that there is another usecase for xsbt.api.Discovery that this PR
does not replace: discovering tests. This is more complicated because
different test frameworks have different ways to discover tests. For
more information, grep for "Fingerprint" in https://github.com/sbt/sbt
and https://github.com/sbt/junit-interface

smarter commented:
Updated, phew! Hopefully this gets merged before the next major reformatting/refactoring of the codebase.
